### PR TITLE
feat: specify share version in `MsgWirePayForBlob`

### PR DESCRIPTION
**Pull request by rootulp**
Closes https://github.com/celestiaorg/celestia-app/issues/936
Opens https://github.com/celestiaorg/celestia-app/issues/1041

### CODE REVIEW

File: pkg/shares/split_sparse_shares.go [hunk 1]
```diff
@@ -22,6 +22,8 @@ func NewSparseShareSplitter() *SparseShareSplitter {
 
 // Write adds the delimited data to the underlying messages shares.
 func (sss *SparseShareSplitter) Write(msg coretypes.Message) {
+	// TODO coretypes.Message likely needs a ShareVersion field so that this
+	// function can contain a switch statement based on it.
```
Comment (rootulp) [inline pkg/shares/split_sparse_shares.go:26]: FWIW this motivated https://github.com/celestiaorg/celestia-app/issues/1022

Comment (rootulp) [reply pkg/shares/split_sparse_shares.go:26]: I see two options:
1. add `ShareVersion` as a field to `coretypes.Blob`. Pro: easy to do, leaves this code unaffected. Con: leaks share version to celestia-core types
2. do something hacky in celestia-app (similar to [ExtractShareIndexes](https://github.com/rootulp/celestia-app/blob/fe097d9630d2d4f8544b3b043c3362f8c6d5b131/pkg/shares/share_splitting.go#L88)) that extracts the share version associated with a blob from it's MsgPayForBlob.

File: x/blob/client/cli/wirepayfordata.go [hunk 2]
```diff
@@ -45,7 +46,10 @@ func CmdWirePayForBlob() *cobra.Command {
 				return fmt.Errorf("failure to decode hex message: %w", err)
 			}
 
-			pfbMsg, err := types.NewWirePayForBlob(namespace, message)
+			// TODO: allow the user to override the share version via a new flag
```
Comment (rootulp) [inline x/blob/client/cli/wirepayfordata.go:49]:
```suggestion
			// TODO: allow the user to override the share version via a new flag
			// See https://github.com/celestiaorg/celestia-app/issues/1041
```

File: go.sum [hunk 3]
```diff
@@ -179,8 +179,8 @@ github.com/btcsuite/winsvc v1.0.0/go.mod h1:jsenWakMcC0zFBFurPLEAyrnc/teJEM1O46f
 github.com/bwesterb/go-ristretto v1.2.0/go.mod h1:fUIoIZaG73pV5biE2Blr2xEzDoMj7NFEuV9ekS419A0=
 github.com/c-bata/go-prompt v0.2.2/go.mod h1:VzqtzE2ksDBcdln8G7mk2RX9QyGjH+OVqOCSiVIqS34=
 github.com/casbin/casbin/v2 v2.1.2/go.mod h1:YcPU1XXisHhLzuxH9coDNf2FbKpjGlbCg3n9yuLkIJQ=
-github.com/celestiaorg/celestia-core v1.8.0-tm-v0.34.20 h1:D1KxHZp8xGksmmYMUNlgMcEq2dnXyGQzU6XBWm7vlh0=
-github.com/celestiaorg/celestia-core v1.8.0-tm-v0.34.20/go.mod h1:2ze96HI7+bV0RxY/EMEInsVJAgYO7sQ1DOt4MnFU/No=
+github.com/celestiaorg/celestia-core v1.9.0-tm-v0.34.20 h1:v+Ay1f/Q4X7+Vo14No6J7f304pObM7aJygwiJ8OSsVc=
+github.com/celestiaorg/celestia-core v1.9.0-tm-v0.34.20/go.mod h1:2ze96HI7+bV0RxY/EMEInsVJAgYO7sQ1DOt4MnFU/No=
```
Comment (evan-forbes) [inline go.sum:183]: :+1:

**Review by evan-forbes — COMMENTED** [review]
thanks for adding this

[comment][no change]
tbh it feels weird that we're adding the option to add specify a share version, but we only support a single share version. I know we still have to add it, but it feels like a potential source of confusion in the future given that only a few people know what it is. 

[comment][no change]
For example, I'm not yet convinced user facing functions, such as celestia-node or our cli, should expose it in their API yet. Perhaps only as an option

**Comment by rootulp** [comment]
> For example, I'm not yet convinced user facing functions, such as celestia-node or our cli, should expose it in their API yet.

FWIW I don't think it needs to be exposed when there only exists one supported share version but I do think it needs to be exposed at the time of introducing another share version format.

> Perhaps only as an optional flag, but maybe not as a required arg

Agreed completely. https://github.com/celestiaorg/celestia-app/issues/1041 intends on making it optional with a sane default (e.g. `0`)

File: testutil/blob/testutil.go [hunk 4]
```diff
@@ -111,8 +113,8 @@ func generateRawWirePFBTx(t *testing.T, txConfig client.TxConfig, ns, message []
 	return rawTx
 }
 
-func generateSignedWirePayForBlob(t *testing.T, ns, message []byte, signer *types.KeyringSigner, options []types.TxBuilderOption) *types.MsgWirePayForBlob {
-	msg, err := types.NewWirePayForBlob(ns, message)
+func generateSignedWirePayForBlob(t *testing.T, ns []byte, message []byte, shareVersion uint8, signer *types.KeyringSigner, options []types.TxBuilderOption) *types.MsgWirePayForBlob {
+	msg, err := types.NewWirePayForBlob(ns, message, shareVersion)
```
[the remark this answers is not in the document]
Comment (rootulp) [reply testutil/blob/testutil.go:117]: Agreed. Modified in this PR to make tests easier to modify when we do support a second `shareVersion` but agreed hard-coding the share version to `0` in tests seems reasonable until there is a second supported share version.

File: x/blob/types/payforblob.go [hunk 5]
```diff
@@ -114,7 +114,7 @@ func BuildPayForBlobTxFromWireTx(
 //
 // [Message layout rationale]: https://github.com/celestiaorg/celestia-specs/blob/e59efd63a2165866584833e91e1cb8a6ed8c8203/src/rationale/message_block_layout.md?plain=1#L12
 // [Non-interactive default rules]: https://github.com/celestiaorg/celestia-specs/blob/e59efd63a2165866584833e91e1cb8a6ed8c8203/src/rationale/message_block_layout.md?plain=1#L36
-func CreateCommitment(namespace, message []byte) ([]byte, error) {
+func CreateCommitment(namespace []byte, message []byte, shareVersion uint8) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (rootulp) [reply x/blob/types/payforblob.go:117]: great catch 🤦‍♂️ this wasn't updated after adding `ShareVersion` to coretypes.Blob. Will fix and add unit tests

[the remark this answers is not in the document]
Comment (rootulp) [reply x/blob/types/payforblob.go:117]: Resolved in https://github.com/celestiaorg/celestia-app/pull/1028/commits/be46b2c3a8e932e83b16c1757e85c3e31763dd33

[the remark this answers is not in the document]
Comment (rootulp) [reply x/blob/types/payforblob.go:117]: Added a test for unsupported share versions in https://github.com/celestiaorg/celestia-app/pull/1028/commits/02cec8bcc2d79f78bcbf0b763743ec482d262380

**Review by evan-forbes — COMMENTED** [review]
lets fix the linter, and then we can merge this. apologies for the late re-review.

this will also change with the upcoming refactor of removing the wirePFB, as we'll have to add it somewhere else (either directly in the PFB, or the BlobTX)

**Comment by rootulp** [comment]
> as we'll have to add it somewhere else (either directly in the PFB, or the BlobTX)

Agreed. This PR adds `share_version` to `MsgPayForBlob` too https://github.com/celestiaorg/celestia-app/pull/1028/files#diff-7ea270de437078b570a910b6909fa9e04b23a253369f18dd1d52faec967eda7aR50 so I expect the removal of `MsgWirePayForBlob` to not cause too many issues with this PR.

**Comment by rootulp** [comment]
> apologies for the late re-review

Absolutely no worries, I think I just fixed all issues so this is ready for re-review

**Comment by rootulp** [comment]
Hold off on merging, while working on an unrelated change I discovered https://github.com/rootulp/celestia-app/blob/26d800681fa9b3ae90ae9b9ad6600dc5e0d5247d/pkg/shares/utils.go#L44-L64 which doesn't account for the newly added share version field

**Review by evan-forbes — APPROVED** [review]
nice!!

**Pull request merged**